library/std/src/sys/unix/ext/net/ancillary.rs

```diff
@@ -147,26 +152,26 @@ fn add_to_ancillary_data<T>(
                           target_os = "emscripten",
                           target_os = "freebsd",
                           all(target_os = "linux", target_env = "musl",),
+                          target_os = "macos",
                           target_os = "netbsd",
                           target_os = "openbsd",
                       ))] {
                 msg.msg_controllen = *length as libc::socklen_t;
             }
         }
 
-        let mut cmsg = libc::CMSG_FIRSTHDR(&msg);
+        let first_cmsg = libc::CMSG_FIRSTHDR(&msg);
+        let mut cmsg = first_cmsg;
         let mut previous_cmsg = cmsg;
         while !cmsg.is_null() {
             previous_cmsg = cmsg;
             cmsg = libc::CMSG_NXTHDR(&msg, cmsg);
-            cfg_if::cfg_if! {
-                // Android return the same pointer if it is the last cmsg.
-                // Therefore, check it if the previous pointer is the same as the current one.
-                if #[cfg(target_os = "android")] {
-                    if cmsg == previous_cmsg {
-                        break;
-                    }
-                }
+
+            // From https://tools.ietf.org/html/rfc2292#section-4.3.2: "if the value of the cmsg
```

Comment:
I tested the previous code on macOS and OpenBSD and it gets stuck in an endless loop, so it does not seem to be correct.

Reply:
Interesting, the original [OpenBSD](https://github.com/openbsd/src/blob/master/sys/sys/socket.h#L496) version does not check if `cmsg` is `NULL`. The [Rust version](https://github.com/rust-lang/libc/blob/master/src/unix/bsd/netbsdlike/openbsd/mod.rs#L1417) of the OpenBSD does. But I do not think that this is the problem because the `CMSG_NXTHDR` is never called with `NULL`. 

If you give me some information I could help you to figured it out. For example, what value has the `cmsg` and `previous_cmsg` pointer in the endless loop?

Reply:
OK, it does return the previous cmsg and not the first one in this case. I will update the PR, thanks for the review!

Reply:
I think the Rust version is basically just a copy + paste from one of the BSD Rust implementations, they all look similar except DragonFly. But you're right, the problem does not seem to be CMSG_NXTHDR. So it's probably in the kernel itself.

What if we just unconditionally break the loop when the pointer is equal to the previous pointer, instead of putting it behind a feature gate, wouldn't that work with Linux as well because it should never happen?

Reply:
Technically, your right, but we should figure out why it does not work in the first place.

Reply:
OK, the problem is related to `cmsg_len`, not `cmsg_controllen`:

1. You zero-initialize `msg`.

2. The Linux code returns NULL when cmsg_len is 0 (or smaller than cmsghdr). https://github.com/rust-lang/libc/blob/master/src/unix/linux_like/linux/mod.rs#L2825

```Rust
        if ((*cmsg).cmsg_len as usize) < ::mem::size_of::<cmsghdr>() {
            return 0 as *mut cmsghdr;
        };
```

2. Android and the BSDs don't do that, so they return the same (previous) pointer when the `cmsg_len` offset is 0. https://github.com/rust-lang/libc/blob/master/src/unix/linux_like/android/mod.rs#L2295

```rust
            (cmsg as usize + _ALIGN((*cmsg).cmsg_len as usize))
                as *mut ::cmsghdr
```

Reply:
Okay, you can remove the feature gate if you want, but I extended the feature. 

In the end thank you for the Bug fix.

Reply:
I checked the implementations in the `libc` crate where Linux and emscripten seem to be the exception, all others including Solaris seem to have the other behavior. I don't think that this simple pointer comparison has a noticeable performance impact, so I preferred to remove the feature gate.

I just pushed an update to the PR, tested it with fd passing on Linux, OpenBSD, and macOS again.

Reply:
@LinkTed if you really prefer it, we could add a feature gate for Linux and Emscripten. What do you think?

```rust
#[cfg(not(any(target_os = "linux", target_os = "emscripten")))]
```

Reply:
It is up to you, I do not have any problem with the current changes.